clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -9,74 +9,78 @@ use rustc_span::symbol::sym;
-                _ => return false,
-            };
-            let inner_span = match get_qpath_generic_tys(qpath).next() {
-                Some(ty) => ty.span,
-                None => return false,
-            };
-            let mut applicability = Applicability::MachineApplicable;
-            span_lint_and_sugg(
-                cx,
-                REDUNDANT_ALLOCATION,
-                hir_ty.span,
-                "usage of `Rc<Box<T>>`",
-                "try",
-                format!(
-                    "Rc<{}>",
-                    snippet_with_applicability(cx, inner_span, "..", &mut applicability)
-                ),
-                applicability,
-            );
-            true
-        } else {
-            utils::match_borrows_parameter(cx, qpath).map_or(false, |span| {
-                let mut applicability = Applicability::MachineApplicable;
-                span_lint_and_sugg(
-                    cx,
-                    REDUNDANT_ALLOCATION,
-                    hir_ty.span,
-                    "usage of `Rc<&T>`",
-                    "try",
-                    snippet_with_applicability(cx, span, "..", &mut applicability).to_string(),
-                    applicability,
-                );
-                true
-            })
-        }
+    let (inner_sym, ty) = if let Some(ty) = is_ty_param_lang_item(cx, qpath, LangItem::OwnedBox) {
+        ("Box", ty)
+    } else if let Some(ty) = is_ty_param_diagnostic_item(cx, qpath, sym::Rc) {
+        ("Rc", ty)
+    } else if let Some(ty) = is_ty_param_diagnostic_item(cx, qpath, sym::Arc) {
+        ("Arc", ty)
+    } else {
+        return false;
+    };
+
+    let inner_qpath = match &ty.kind {
+        TyKind::Path(inner_qpath) => inner_qpath,
+        _ => return false,
+    };
+    let inner_span = match get_qpath_generic_tys(inner_qpath).next() {
+        Some(ty) => ty.span,
+        None => return false,
+    };
+    let mut applicability = Applicability::MaybeIncorrect;
+    if inner_sym == outer_sym {
+        span_lint_and_sugg(
+            cx,
+            REDUNDANT_ALLOCATION,
+            hir_ty.span,
+            &format!("usage of `{}<{}<T>>`", outer_sym, inner_sym),
+            "try",
+            format!(
+                "{}<{}>",
+                outer_sym,
+                snippet_with_applicability(cx, inner_span, "..", &mut applicability)
+            ),
+            applicability,
+        );
     } else {
-        false
+        span_lint_and_help(
+            cx,
+            REDUNDANT_ALLOCATION,
+            hir_ty.span,
+            &format!(
+                "you seem to be trying to use `{}<{}<T>>`. Consider using just `{}<T>` of `{}<T>`",
+                outer_sym, inner_sym, outer_sym, inner_sym,
+            ),
+            None,
+            &format!(
+                "`{}<T>` is already on the heap, `{}<{}<T>>` makes an extra allocation",
+                inner_sym, outer_sym, inner_sym,
+            ),
+        );
```

Comment:
The lint message already contains a help message. It would be better to add that as a help message. You can use `span_lint_and_then(..)` to add two help messages. I would also change the current "help" message to be a "note" as it's more an explanation.

Reply:
Sorry for making so many mistakes.

Reply:
It's totally fine, don't worry about it. They are all pretty minor and some, were actually just NITs from the previous implementation. I for my part had fun reviewing this PR and to see it take shape, and I'm also happy that you're still motivated :upside_down_face:.

Reply:
I' m not sure about note information. If this `span_lint_and_note` is ok?

Reply:
You could in theory use `span_lint_and_note` but that would only give you the ability to add a single note to a diagnostic. `span_lint_and_then` enables you to add several messages.

In this case, it can look something like this (Haven't tested it locally, but it should work with some slight adjustments).
```rs
span_lint_and_then(
    cx,
    REDUNDANT_ALLOCATION,
    hir_ty.span,
    &format!("use of `{}<{}<T>>`", outer_sym, inner_sym),
    |diag| {
        diag.note(&format!(
            "`{}<T>` is already on the heap, `{}<{}<T>>` makes an extra allocation",
            inner_sym, outer_sym, inner_sym,
        ));
        diag.help(&format!("consider using just `{}<T>` or `{}<T>`", outer_sym, inner_sym));
    },
);
```

That should give you a message like:

```
error: use of `Box<Arc<T>>`.
  --> $DIR/redundant_allocation.rs:33:30
   |
LL |     pub fn box_test9<T>(foo: Box<Arc<T>>) -> Box<Arc<SubT<T>>> {
   |                              ^^^^^^^^^^^
   |
   = note: `Arc<T>` is already on the heap, `Box<Arc<T>>` makes an extra allocation
   = help: consider using just `Box<T>` or `Arc<T>`
```

Other links:
* [Example from inside Clippy](https://github.com/rust-lang/rust-clippy/blob/12fce557669a0de230399cf8e6eee4f5307bf87b/clippy_lints/src/fallible_impl_from.rs#L130-L140)
* [DiagnosticBuilder docs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.note)

Reply:
I found the note information is different among `Box<Box<T>>` and `Box<Rc<T>>`.
So, in order to add a note to `span_lint_and_sugg`, I change it into `span_lint_and_then`.